Avoid allocations of reusable CanBuildFroms

mkeskells opened the pull request:
(For 2.12.x's eyes only.)

Use vals to cache a single instance of stateless CanBuildFrom instances.
These are cast by the existing `implicit def` to the suitable generic
type. This pattern was already used in some places -- this PR applies
it systematically across `collection.{mutable,immutable}`.

The `CanBuildFrom` instances for arrays and wrapped arrays are
cached for each primitive type, Unit, and Object. Each of these
instances is backed by a dedicated subclass of CanBuildFrom that
avoids subsequent dispatch on the `ClassTag[T]`.

hrhino commented on src/library/scala/collection/mutable/WrappedArray.scala:
```diff
@@ -114,13 +114,44 @@ object WrappedArray {
+      case java.lang.Boolean.TYPE   => cbfBooleanArray
+      case java.lang.Void.TYPE      => cbfUnitArray
+      case objectClass              => cbfObjectArray
+      case _ => makeCBF[T with AnyRef]
```
Why `with AnyRef`? I think that's why the bootstrapped compilation is failing.

mkeskells replied:
seemed implied in IDE. Removed and updated objectClass to ObjectClass

hrhino commented on src/library/scala/Array.scala:
```diff
@@ -61,7 +61,35 @@ object Array extends FallbackArrayBuilding {
+      case java.lang.Boolean.TYPE   => cbfBooleanArray
+      case java.lang.Void.TYPE      => cbfUnitArray
+      case objectClass              => cbfObjectArray
+      case _ => makeCBF[T with AnyRef]
```
Ditto.

mkeskells replied:
seemed implied in IDE. Removed and updated objectClass to ObjectClass

hrhino commented on src/library/scala/collection/TraversableView.scala:
```diff
@@ -25,6 +25,10 @@ trait TraversableView[+A, +Coll] extends TraversableViewLike[A, Coll, Traversabl
 object TraversableView {
+  private[this] val reusedNoBuilder = new NoBuilder[Any]
+  private[collection] def noBuilder[A] = reusedNoBuilder.asInstanceOf[NoBuilder[A]]
+  @deprecated("use noBuilder")
```
Seems weird to deprecate it when you can't use the replacement.

mkeskells replied:
maybe at a later stage.
removed for now

mkeskells replied:
removed for now

hrhino commented on src/library/scala/collection/TraversableView.scala:
```diff
@@ -25,6 +25,10 @@ trait TraversableView[+A, +Coll] extends TraversableViewLike[A, Coll, Traversabl
  */
 object TraversableView {
+  private[this] val reusedNoBuilder = new NoBuilder[Any]
+  private[collection] def noBuilder[A] = reusedNoBuilder.asInstanceOf[NoBuilder[A]]
```
I'm pretty sure this will need a MiMa exception.

sjrd commented on src/library/scala/Array.scala:
```diff
@@ -61,7 +61,35 @@ object Array extends FallbackArrayBuilding {
-  implicit def canBuildFrom[T](implicit t: ClassTag[T]): CanBuildFrom[Array[_], T, Array[T]] =
+  implicit def canBuildFrom[T](implicit t: ClassTag[T]): CanBuildFrom[Array[_], T, Array[T]] = {
+    val tag = implicitly[ClassTag[T]]
+    (tag.runtimeClass match {
```
Is this match really an improvement on the JVM? Because at least on JS this is going to be detrimental.

mkeskells replied:
its a saving on the allocation of the CanBuildFrom, not the CPU

mkeskells replied:
previously a use of a CBF would switch on the runtime class twice for each array, this changes it it 3 times. If its really an issue then we can specialise the code for each case and reduce to calls to one for each usage

what is the measured effect on JS? If so its already an issue

sjrd replied:
Well in JS de can entirely stack allocate the `CanBuildFrom` and the `ArrayBuilder` down to completely eliminating them and constant-folding the `ClassTag`s. We can do that because they are new instances. If they first switch to fetch an existing instance on the heap, we don't know what's in the heap and we can't constant-fold its members.

But I realized that we already override `Array.scala` with our own implementation to achieve this. So I guess the change in this PR won't affect us. At least not beyond the fact that it widens the gap between the JVM and the JS implementation of the library.

retronym replied:
Also, on the JVM, it is best to first switch on `tag.runtimeClass.isPrimitive` to have a fast path for non-primitive cases.

retronym replied:
@sjrd I'd be open to keeping an `private[scala] final val isScalaJS = false` constant in the library if that would help you avoid copy/paste. You could recompile with that constant flipped and get the original code here, for example.

mkeskells replied:
I did consider the isPrimitive check, but wanted to discuss. 
Will rework, but it should also apply in other cases. There seem to be 9 locations to apply this

mkeskells replied:
I see that @retronym fixed these

sjrd commented on src/library/scala/collection/mutable/WrappedArray.scala:
```diff
@@ -114,13 +114,44 @@ object WrappedArray {
-  implicit def canBuildFrom[T](implicit m: ClassTag[T]): CanBuildFrom[WrappedArray[_], T, WrappedArray[T]] =
+  implicit def canBuildFrom[T](implicit m: ClassTag[T]): CanBuildFrom[WrappedArray[_], T, WrappedArray[T]] = {
+    val tag = implicitly[ClassTag[T]]
+    (tag.runtimeClass match {
```
Same question: is this really an improvement?

mkeskells replied:
same answer

retronym commented on src/library/scala/Array.scala:
```diff
@@ -61,7 +61,35 @@ object Array extends FallbackArrayBuilding {
+  private[this] val cbfLongArray    = makeCBF[Long]
+  private[this] val cbfShortArray   = makeCBF[Short]
+  private[this] val cbfUnitArray    = makeCBF[Unit]
+  private[this] val cbfObjectArray  = makeCBF[Object]
```
If we're going this far, we might as well make a dedicated subclass for each of these that avoids the indirection through another type test in `ArrayBuilder.make`.

Both these suggestions also apply to `WrappedArray`.

mkeskells replied:
will do

retronym replied:
I've pushed this change myself.

retronym commented on src/library/scala/collection/mutable/HashSet.scala:
```diff
@@ -104,7 +104,9 @@ extends AbstractSet[A]
-  implicit def canBuildFrom[A]: CanBuildFrom[Coll, A, HashSet[A]] = setCanBuildFrom[A]
+  implicit def canBuildFrom[A]: CanBuildFrom[Coll, A, HashSet[A]] =
+    ReusableCBF.asInstanceOf[CanBuildFrom[Coll, A, HashSet[A]]]
+  private[this] val ReusableCBF = setCanBuildFrom[Any]
```
How about we make `GenSetFactory.setCanBuildFrom` backed by a val, and leave this as is? There might be a few other places where we the same approach works.

mkeskells replied:
doesnt it need to be able to refer to newBuilder etc in this scope?

retronym replied:
Yes, I misunderstood the code.

hrhino commented on src/reflect/mima-filters/2.12.0.forwards.excludes:
```diff
@@ -44,6 +44,8 @@ ProblemFilters.exclude[IncompatibleMethTypeProblem]("scala.reflect.io.ZipArchive
 ProblemFilters.exclude[IncompatibleResultTypeProblem]("scala.reflect.io.FileZipArchive.allDirs")
 ProblemFilters.exclude[DirectMissingMethodProblem]("scala.reflect.runtime.SynchronizedSymbols#SynchronizedSymbol.typeConstructor")
 
+ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.TraversableView.noBuilder")
```
_pace_ MiMa, this might need to be `scala.collection.TraversableView$.noBuilder`.

retronym replied:
It's actually just in the wrong file, should be in `src/library/...`. I'll push a fix.

retronym committed:
[nomerge] Avoid allocations of reusable CanBuildFroms

(For 2.12.x's eyes only.)

Use vals to cache a single instance of stateless CanBuildFrom instances.
These are cast by the existing `implicit def` to the suitable generic
type. This pattern was already used in some places -- this PR applies
it systematically across `collection.{mutable,immutable}`.

The `CanBuildFrom` instances for arrays and wrapped arrays are
cached for each primitive type, Unit, and Object. Each of these
instances is backed by a dedicated subclass of CanBuildFrom that
avoids subsequent dispatch on the `ClassTag[T]`.

retronym reviewed (approved):
Approving of the idea. We should get another set of 👀 to review this as I was involved in the code change as well.

hrhino reviewed (approved):
Looks great! I'd like Stefan's opinion, too.

mkeskells commented:
@retronym thanks for progressing this PR ⭐️

hrhino commented on src/library/scala/Array.scala:
```diff
@@ -61,10 +61,79 @@ object Array extends FallbackArrayBuilding {
 
-  implicit def canBuildFrom[T](implicit t: ClassTag[T]): CanBuildFrom[Array[_], T, Array[T]] =
+  implicit def canBuildFrom[T](implicit t: ClassTag[T]): CanBuildFrom[Array[_], T, Array[T]] = {
+    val tag = implicitly[ClassTag[T]]
```
a.k.a. `t` but hey

mkeskells committed:
use existing ClassTag

diesalbla commented on src/library/scala/reflect/ClassTag.scala:
```diff
@@ -59,17 +59,20 @@ trait ClassTag[T] extends ClassManifestDeprecatedApis[T] with Equals with Serial
-      case java.lang.Boolean.TYPE   => new Array[Boolean](len).asInstanceOf[Array[T]]
-      case java.lang.Void.TYPE      => new Array[Unit](len).asInstanceOf[Array[T]]
-      case _                        => java.lang.reflect.Array.newInstance(runtimeClass, len).asInstanceOf[Array[T]]
+    if (runtimeClass.isPrimitive) {
```
```suggestion
    if (runtimeClass.isPrimitive) runtimeClass match {
```
This may reduce some whitespace diff noise.

diesalbla commented on src/library/scala/collection/mutable/WrappedArrayBuilder.scala:
```diff
@@ -36,17 +36,20 @@ class WrappedArrayBuilder[A](tag: ClassTag[A]) extends ReusableBuilder[A, Wrappe
-      case java.lang.Boolean.TYPE   => new WrappedArray.ofBoolean(new Array[Boolean](size)).asInstanceOf[WrappedArray[A]]
-      case java.lang.Void.TYPE      => new WrappedArray.ofUnit(new Array[Unit](size)).asInstanceOf[WrappedArray[A]]
-      case _                        => new WrappedArray.ofRef[A with AnyRef](tag.newArray(size).asInstanceOf[Array[A with AnyRef]]).asInstanceOf[WrappedArray[A]]
+    val newelems = if (runtimeClass.isPrimitive) {
```
```suggestion
    if (runtimeClass.isPrimitive) runtimeClass match {
```
This may reduce some whitespace diff noise.

diesalbla commented on src/library/scala/collection/mutable/ArrayBuilder.scala:
```diff
@@ -37,17 +37,21 @@ object ArrayBuilder {
-      case java.lang.Void.TYPE      => new ArrayBuilder.ofUnit().asInstanceOf[ArrayBuilder[T]]
-      case _                        => new ArrayBuilder.ofRef[T with AnyRef]()(tag.asInstanceOf[ClassTag[T with AnyRef]]).asInstanceOf[ArrayBuilder[T]]
+    val cls = tag.runtimeClass
+    if (cls.isPrimitive) {
```
```suggestion
    if (cls.isPrimitive) cls match {
```
This may reduce some whitespace diff noise.

retronym commented:
paging @szeiger for a look at this one. I'm happy with the PR myself.

szeiger commented on src/library/scala/Array.scala:
```diff
@@ -61,10 +61,78 @@ object Array extends FallbackArrayBuilding {
+  implicit def canBuildFrom[T](implicit tag: ClassTag[T]): CanBuildFrom[Array[_], T, Array[T]] = {
+    val cls = tag.runtimeClass
+    (if (cls.isPrimitive) {
+      cls match {
```
This gets compiled to a cascade of ifs, so order matters. I'd put frequently used primitive types (probably Int, Long, Double, the same ones for which streams are optimized) first.

mkeskells replied:
changed this and WrappedArray order to match `def make` in `WrappedArray`

szeiger commented on src/library/scala/collection/BitSet.scala:
```diff
@@ -32,6 +32,7 @@ object BitSet extends BitSetFactory[BitSet] {
   /** $canBuildFromInfo */
-  implicit def canBuildFrom: CanBuildFrom[BitSet, Int, BitSet] = bitsetCanBuildFrom
+  implicit def canBuildFrom: CanBuildFrom[BitSet, Int, BitSet] = ReusableCBF
+  private[this] val ReusableCBF = bitsetCanBuildFrom
```
Could we just use an `implicit val` here? You shouldn't need an extra method unless it's polymorphic.

mkeskells replied:
will do

mkeskells replied:
done

szeiger commented on src/library/scala/collection/TraversableView.scala:
```diff
@@ -25,6 +25,9 @@ trait TraversableView[+A, +Coll] extends TraversableViewLike[A, Coll, Traversabl
  *  `TraversableView`s work. Its definitions are generally not accessed directly by clients.
  */
 object TraversableView {
+  private[this] val reusedNoBuilder = new NoBuilder[Any]
```
What do we gain from caching this value? You can't do anything useful with `NoBuilder`. Is it ever created but then not used?

mkeskells replied:
I was just following the same pattern - reverting the changes to this file

mkeskells replied:
done

szeiger commented on src/library/scala/collection/IterableView.scala:
```diff
@@ -27,8 +27,9 @@ trait IterableView[+A, +Coll] extends IterableViewLike[A, Coll, IterableView[A,
+    ReusableCBF.asInstanceOf[CanBuildFrom[Coll, A, IterableView[A, Iterable[_]]]]
+  private[this] val ReusableCBF = new CanBuildFrom[Coll, Any, IterableView[Any, Iterable[_]]] {
+      def apply(from: Coll) = noBuilder[Any]
+      def apply() = noBuilder
```
(See below.) If it always returns `NoBuilder` are there any important cases where it is created in the first place?

mkeskells replied:
I was just following the same pattern - reverting the changes to this file

szeiger commented on src/library/scala/collection/mutable/WrappedArray.scala:
```diff
@@ -114,13 +114,75 @@ object WrappedArray {
+
+  private[this] val ObjectClass = classOf[Object]
+
+  private[this] val cbfBooleanArray = new CanBuildFrom[WrappedArray[_], Boolean, WrappedArray[Boolean]] {
+    def apply(from: WrappedArray[_]) = new ArrayBuilder.ofBoolean mapResult WrappedArray.make[Boolean]
+    def apply = new ArrayBuilder.ofBoolean mapResult WrappedArray.make[Boolean]
+  }
```
It would be nice to have `CanBuildFrom:mapResult` but I guess it's too late for that now.

mkeskells replied:
maybe in 2.14?

mkeskells committed:
review comments
order matches in expected frequency order (Array, WrappedArray + associated builders, and ClassTag.newArray)
avoid extra def in BitSets
don't optimise for NoBuilder cases

szeiger commented on src/library/scala/collection/immutable/BitSet.scala:
```diff
@@ -81,7 +81,8 @@ object BitSet extends BitSetFactory[BitSet] {
   /** $bitsetCanBuildFrom */
-  implicit def canBuildFrom: CanBuildFrom[BitSet, Int, BitSet] = bitsetCanBuildFrom
+  implicit def canBuildFrom: CanBuildFrom[BitSet, Int, BitSet] = ReusableCBF
+  private[this] val ReusableCBF = bitsetCanBuildFrom
```
Here's another case where we don't need a new `val`. All monomorphic ones can use `val canBuildFrom`.

mkeskells replied:
done

szeiger commented on src/library/scala/collection/immutable/WrappedString.scala:
```diff
@@ -56,7 +56,9 @@ final class WrappedString(val self: String) extends AbstractSeq[Char] with Index
+  implicit def canBuildFrom: CanBuildFrom[WrappedString, Char, WrappedString] =
+    ReusableCBF.asInstanceOf[CanBuildFrom[WrappedString, Char, WrappedString]]
+  private[this] val ReusableCBF = new CanBuildFrom[WrappedString, Char, WrappedString] {
     def apply(from: WrappedString) = newBuilder
```
This one, too

mkeskells replied:
done

szeiger commented on src/library/scala/collection/mutable/BitSet.scala:
```diff
@@ -193,7 +193,8 @@ object BitSet extends BitSetFactory[BitSet] {
   /** $bitsetCanBuildFrom */
-  implicit def canBuildFrom: CanBuildFrom[BitSet, Int, BitSet] = bitsetCanBuildFrom
+  implicit def canBuildFrom: CanBuildFrom[BitSet, Int, BitSet] = ReusableCBF
+  private[this] val ReusableCBF = bitsetCanBuildFrom
```
another one

mkeskells replied:
done

szeiger reviewed (approved):
Other than the remaining unnecessary forwarders this looks good.

mkeskells committed:
review comments
use direct vals where no casting is required
